woocommerce/woocommerce-gutenberg-products-block
# Hide payment methods that have missing dependencies and display an error in the admin

Fixes #3800.

In some occasions, a payment method might be registered but its dependencies might not (read #3800 for a more in-depth explanation). In those cases, this PR makes it so:
* The payment method with missing dependencies is removed from Cart/Checkout script dependencies. This way, Cart and Checkout block continue working and the broken payment method is silently hidden.
* It adds an error notice in `wp-admin` and `error_log()`'s the error. If `WP_DEBUG` is set to true, it also includes a link to our docs.

### How to test the changes in this Pull Request:

There are two different set-ups that you can try:
* One consists on recreating the issue we faced with WC Pay. You will need to install WC Pay and remove [these lines](https://github.com/Automattic/woocommerce-payments/blob/trunk/includes/class-wc-payments-blocks-payment-method.php#L37-L43), so `stripe` is not registered.
* Another solution is to add a fake dependency in any other payment method. Ie, in [Cheque](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/Payments/Integrations/Cheque.php):
```diff
	public function get_payment_method_script_handles() {
		$this->asset_api->register_script(
			'wc-payment-method-cheque',
			'build/wc-payment-method-cheque.js',
+			[ 'missing-dependency' ]
		);
		return [ 'wc-payment-method-cheque' ];
	}
```

1. Go to any `wp-admin` page and verify an error notice is displayed for each faulty payment method:
![imatge](https://user-images.githubusercontent.com/3616980/109959632-0e5a7c00-7ce8-11eb-8ebe-b3d9b4233850.png)
2. Create a post or page and verify the Cart and Checkout blocks are available. If you add them, you will notice the faulty payment methods are not available.
3. In the frontend, verify the Cart and Checkout blocks load as usual (but the payment methods with missing dependencies are not loaded).

### Changelog

> Prevent errant payment methods from keeping Cart and Checkout blocks from loading
